velox/functions/lib/LikeFunctionUtils.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/lib/LikeFunctionUtils.h"
+#include "velox/vector/TypeAliases.h"
+
+namespace facebook::velox::functions {
+namespace likeutils {
+
+bool isWildcardInPattern(const std::string& pattern) {
+  vector_size_t patternLength = pattern.size();
+  for (vector_size_t i = 0; i < patternLength; i++) {
+    if ((pattern[i] == '%') || (pattern[i] == '_')) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool matchExactPattern(const std::string& input, const std::string& pattern) {
+  vector_size_t i = 0;
+  vector_size_t j = 0;
+  vector_size_t inputLength = input.size();
+  vector_size_t patternLength = pattern.size();
+
+  if (inputLength != patternLength) {
+    return false;
+  }
+
+  for (; (i < inputLength) && (j < patternLength); i++, j++) {
+    if (input[i] != pattern[j]) {
+      return false;
+    }
+  }
+  return ((i == inputLength) && (j == patternLength));
+}
+
+bool isPrefixPattern(const std::string& pattern) {
+  vector_size_t patternLength = pattern.size();
+  vector_size_t i = patternLength - 1;
+
+  for (; i >= 0; i--) {
```

Comment:
This loop is repeated for each input string unnecessarily. We can compute the length of the prefix once, then use 'memcmp' function to compare the prefixes of input and pattern. Same for suffix matching.

Reply:
Could you please share how this loop can be simplified? I have simplified the matching function using memcmp and am using the lengths of prefix/suffix that would be precalculated here, as suggested. I am not sure how prefix/suffix check functions can be further simplified.

Reply:
Before processing any data, we can determine the kind of pattern. If it is a prefix, we can further calculate the length of the prefix. When processing data, we can simply memcpy first N bytes of the data to match first N bytes of the pattern. We don't need to compute N for each input string.

Reply:
Thanks for clarifying, the current code does the same. In the constructor, the kind of pattern is checked and this state is saved. If the pattern is a prefix/suffix, the prefix/suffix length is also saved. memcmp is then used on the number of characters to be matched as per the pattern kind.